### PR TITLE
[Seq] Add `seq.hlmem` ODS ops and types

### PR DESCRIPTION
**Pull request by mortbopet**

### CODE REVIEW

**Review by teqdruid — CHANGES_REQUESTED** [review]
See all my comments in the original ods + rationale pr.

**Comment by mortbopet** [comment]
@teqdruid Changes:
* `seq.hlmem` now defines a single value of `!seq.hlmem<shape x element type>` which is referenced by the port ops.
* `seq.hlmem` now only defines the memory dimension and element type.
* I went ahead and added "support" for multidimensional memories (this was mostly due to laziness that it wasn't present in the original commit). I think this is a reasonable addition:
  * lessens the burden of producers of this commit to themselves implement memory flattening (for whatever abstraction they're generating from)
  * memory flattening should be simple to reason about in the context of `seq.hlmem` (we now its a physical memory).
  * If someone actually needs multidimensional memories in their memory lowering, this allows for it. I initially only plan on writing a unidimensional memory lowering, but the whole point of this op is that it captures a set of fairly generic requirements for a memory implementation, and then it's up to the user to ensure that there exists a memory implementation pass which actually matches the requested configuration
* Latencies are now a per-port property.

**Comment by mortbopet** [comment]
@teqdruid I'm now considering whether we should also add `[read/write]Finished` results for ports. It's fairly trivial to implement these in a `hlmem` lowering, an likewise annoying for users of this op to manually have to keep track of this (latency sensitive users obv. still have to design around this, but i don't expect there to be any overhead in including this in the port lowering. if there's no user of the `finished` signals, all logic associated with it (upstream regs) should be optimized away anyways).

**Comment by teqdruid** [comment]
RE the finished signals: I'm like 90% sure the only dialects which could make use of this signal would be handshake/esi. Non dynamic dataflow dialects need to know the latency and schedule around it -- including most HLS technologies. So I don't think it belongs in the `seq` dialect. I think a wrapper around this in a higher level dialect is more appropriate.

File: include/circt/Dialect/Seq/SeqOps.td [hunk 1]
```diff
@@ -115,3 +115,47 @@ def FirRegOp : SeqOp<"firreg",
...
+  [DeclareOpInterfaceMethods<OpAsmOpInterface, ["getAsmResultNames"]>]> {
+  let summary = "Structural read access to a seq.hlmem";
+  let arguments = (ins 
+    HLMemType:$memory,
+    Variadic<AnyType>:$addresses,
+    I1:$rdEn,
+    I64Attr:$latency
+  );
+  let results = (outs AnyType:$readData);
+  let hasCustomAssemblyFormat = 1;
```
Comment (teqdruid) [inline include/circt/Dialect/Seq/SeqOps.td:146]: Could this be assembly format with `custom` segments? May or may not make sense.

File: docs/Dialects/Seq/RationaleSeq.md [hunk 2]
```diff
@@ -226,34 +226,27 @@ merging, read/write conflicts, etc.) and may serve as a target for other
...
-level of abstraction as the core RTL dialects and  contain no notion of control
-flow.
+- Memory *allocation* is defined by the `seq.hlmem` operation. This operation
+defines the internal memory structure. For now, this strictly pertains to the
+layout of the memory (dimensionality) and element type.
+- Memory *access* is defined by separate port operations which reference the
+allocated memory. Port access operations are defined at the same level of
+abstraction as the core RTL dialects and contain no notion of control
+flow. As such, for e.g. a write port with a non-zero latency, the encapsulating
+IR must already have accounted for this latency.
```
Comment (teqdruid) [inline docs/Dialects/Seq/RationaleSeq.md:236]: Should we add a note about conflicting writes? 'The behavior of conflicting writes is defined by the lowering. Generally speaking, it should be considered as undefined.'

**Review by teqdruid — APPROVED** [review]
Looks good. Thanks!

File: include/circt/Dialect/Seq/SeqOps.td (hunk 1 above)
Comment (mortbopet) [reply include/circt/Dialect/Seq/SeqOps.td:146]: I don't think there's any parts of this which are actually complex to parse (warranting `custom`). The main issue with using the ODS format here is the variadic addresses which have type constraints on each value wrt. the width of the corresponding dimension of the memory. This is ''neatly'' handled by `memref` in that addresses are of `index` type (nothing to check).

File: include/circt/Dialect/Seq/SeqOps.td [hunk 3]
```diff
@@ -115,3 +115,47 @@ def FirRegOp : SeqOp<"firreg",
...
+}
+
+
+def WritePortOp : SeqOp<"write", []> {
+  let summary = "Structural write access to a seq.hlmem";
+  let arguments = (ins
+    HLMemType:$memory,
+    Variadic<AnyType>:$addresses,
+    AnyType:$inData,
```
[the remark this answers is not in the document]
Comment (mortbopet) [reply include/circt/Dialect/Seq/SeqOps.td:155]: Added.

File: include/circt/Dialect/Seq/SeqOps.td [hunk 4]
```diff
@@ -115,3 +115,47 @@ def FirRegOp : SeqOp<"firreg",
...
+  let results = (outs AnyType:$readData);
+  let hasCustomAssemblyFormat = 1;
+}
+
+
+def WritePortOp : SeqOp<"write", []> {
+  let summary = "Structural write access to a seq.hlmem";
+  let arguments = (ins
+    HLMemType:$memory,
+    Variadic<AnyType>:$addresses,
```
[the remark this answers is not in the document]
Comment (mortbopet) [reply include/circt/Dialect/Seq/SeqOps.td:154]: added ODS predicate to check # of address values provided as well as address bit widths.

**Pull request merged**